src/libraries/Microsoft.PowerFx.Interpreter/UDFHelper.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Interpreter
+{
+    internal class UDFHelper
+    {
+        public class LazyCheck
+        {
+            internal string _expressionText;
+            internal RecordType _parameterType;
+            internal ParserOptions _options;
+            internal Engine _engine;
+
+            public LazyCheck(Engine engine, string expressionText, RecordType parameterType = null, ParserOptions options = null)
+            {
+                _engine = engine;
+                _expressionText = expressionText;
+                _parameterType = parameterType;
+                _options = options;
+            }
+
+            public CheckResult Get() => _engine.Check(_expressionText, _parameterType, _options);
+        }
+
+        // would also be a good generic helper 
+        private static NamedValue[] Zip(NamedFormulaType[] parameters, FormulaValue[] args)
+        {
+            if (parameters.Length != args.Length)
+            {
+                throw new ArgumentException();
+            }
+
+            var result = new NamedValue[args.Length];
+            for (var i = 0; i < args.Length; i++)
+            {
+                result[i] = new NamedValue(parameters[i].Name, args[i]);
+            }
+
+            return result;
+        }
+        
+        public class UDFDefinition
```

Comment:
nits: Each class in here should be a separate file, and not inner classes of UDFHelper

Reply:
Done.